pkg/servicemesh/control-plane.go

```diff
@@ -0,0 +1,578 @@
+					Storage:          &elasticsearchStorage,
+					RedundancyPolicy: elasticsearchRedundancyPolicy,
+					IndexCleaner:     &elasticsearchIndexCleaner,
+				},
+			},
+			Ingress: &istioV2.JaegerIngressConfig{
+				Enablement: istioV2.Enablement{
+					Enabled: &ingressEnablement,
+				},
+				Metadata: &istioV2.MetadataConfig{
+					Labels:      ingressLabels,
+					Annotations: ingressAnnotations,
+				},
+			},
+		},
+	}
+
+	if builder.Definition.Spec.Addons == nil {
+		builder.Definition.Spec.Addons = &istioV2.AddonsConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Jaeger == nil {
+		builder.Definition.Spec.Addons.Jaeger = &istioV2.JaegerAddonConfig{}
+	}
+
+	builder.Definition.Spec.Addons.Jaeger = jaegerAddon
+
+	return builder
+}
+
+// WithKialiAddon add kiali addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithKialiAddon(
+	enablement bool,
+	name string,
+	dashboardViewOnly bool,
+	dashboardEnableGrafana bool,
+	dashboardEnablePrometheus bool,
+	dashboardEnableTracing bool,
+	serviceLabels map[string]string,
+	serviceAnnotations map[string]string,
+	serviceNodePort int32,
+	ingressEnablement bool,
+	ingressLabels map[string]string,
+	ingressAnnotations map[string]string,
+	ingressHosts []string,
+	ingressContextPath string,
+	ingressTLS istioV1.HelmValues,
+) *ControlPlaneBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the Kiali addons defined", builder.Definition.Name)
+
+	kialiAddon := &istioV2.KialiAddonConfig{
+		Enablement: istioV2.Enablement{
+			Enabled: &enablement,
+		},
+		Name: name,
+		Install: &istioV2.KialiInstallConfig{
+			Dashboard: &istioV2.KialiDashboardConfig{
+				ViewOnly:         &dashboardViewOnly,
+				EnableGrafana:    &dashboardEnableGrafana,
+				EnablePrometheus: &dashboardEnablePrometheus,
+				EnableTracing:    &dashboardEnableTracing,
+			},
+			Service: &istioV2.ComponentServiceConfig{
+				Metadata: &istioV2.MetadataConfig{
+					Labels:      serviceLabels,
+					Annotations: serviceAnnotations,
+				},
+				NodePort: &serviceNodePort,
+				Ingress: &istioV2.ComponentIngressConfig{
+					Enablement: istioV2.Enablement{
+						Enabled: &ingressEnablement,
+					},
+					Metadata: &istioV2.MetadataConfig{
+						Labels:      ingressLabels,
+						Annotations: ingressAnnotations,
+					},
+					Hosts:       ingressHosts,
+					ContextPath: ingressContextPath,
+					TLS:         &ingressTLS,
+				},
+			},
+		},
+	}
+
+	if builder.Definition.Spec.Addons == nil {
+		builder.Definition.Spec.Addons = &istioV2.AddonsConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Kiali == nil {
+		builder.Definition.Spec.Addons.Kiali = &istioV2.KialiAddonConfig{}
+	}
+
+	builder.Definition.Spec.Addons.Kiali = kialiAddon
+
+	return builder
+}
+
+// WithPrometheusAddon add prometheus addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithPrometheusAddon(
```

Comment:
Can you pass:
install *istioV2.PrometheusAddonConfig here like you did in the WithGrafanaAddon function?
It would reduce the number of arguments dramatically and the only verification you'll need to run is:

	if PrometheusAddonConfig == nil {
		glog.V(100).Infof("The PrometheusAddonConfig of ServiceMeshControlPlane can not be empty")

		builder.errorMsg = "'PrometheusAddonConfig' can not be empty for ServiceMeshControlPlane struct"
	}

You can built PrometheusAddonConfig in your test cases for now and later we will implement prometheusaddonconfig.go struct in this package

Reply:
please check my comment: https://github.com/openshift-kni/eco-goinfra/pull/147/files#r1362954188